lib: added logger api to node core

mertcanaltin opened the pull request:
## Description

Adds an experimental `node:logger` module that provides a structured, high-performance logging API for Node.js. Available behind the `--experimental-logger` flag.

Refs: https://github.com/nodejs/node/issues/49296

### Architecture

- Uses `diagnostics_channel` as the dispatch mechanism, loggers publish log records to level-specific channels, consumers subscribe to receive them
- Pre-serializes bindings at construction time to avoid per-log overhead
- Disabled log levels become `noop` functions, eliminating runtime level checks
- JSON output uses the existing internal `Utf8Stream` (SonicBoom port) for high-throughput writes

### API Surface

- **`Logger`**, Producer class with `trace`, `debug`, `info`, `warn`, `error`, `fatal` methods
- **`LogConsumer`**, Base consumer class for custom log handling
- **`JSONConsumer`**, Built-in consumer that outputs structured JSON via `Utf8Stream`
- **`child(bindings, options)`**, Child loggers with inherited context and serializers
- **`logger.<level>.enabled`**, Getter for conditional logging (typo-safe)
- **`stdSerializers`**, Built-in serializers for `err`, `req`, `res`
- **`serialize` symbol**, `Symbol.for('nodejs.logger.serialize')`, custom object serialization hook (stacked with field serializers)

Log levels follow RFC 5424 numerical ordering internally (`trace: 10, debug: 20, info: 30, warn: 40, error: 50, fatal: 60`). The API accepts string level names. `diagnostics_channel` channel names (`log:trace`, `log:debug`, `log:info`, `log:warn`, `log:error`, `log:fatal`) are used internally and can be subscribed to directly via `diagnostics_channel` if needed.

### Performance

Benchmarked against Pino (output to `/dev/null`, n=100000):

| Scenario | node:logger | pino | Difference |
|----------|------------|------|------------|
| simple | 6.06M ops/s | 3.48M ops/s | +74% |
| child | 5.76M ops/s | 4.41M ops/s | +31% |
| disabled | 174M ops/s | 146M ops/s | +19% |
| fields | 2.13M ops/s | 1.36M ops/s | +56% |

### New files

- `lib/logger.js`, Main module (`Logger`, `LogConsumer`, `JSONConsumer`)
- `lib/internal/logger/serializers.js`, Standard `err`/`req`/`res` serializers
- `doc/api/logger.md`, Full API documentation
- `benchmark/logger/basic-json.js`, Benchmarks
- `test/parallel/test-log-basic.js`, Core logger tests
- `test/parallel/test-logger-serializers.js`, Serializer tests

### Modified files

- `src/node_options.h` / `src/node_options.cc`, `--experimental-logger` flag
- `lib/internal/process/pre_execution.js`, `setupLogger()` to allow module loading when flag is set
- `lib/internal/bootstrap/realm.js`, Module registration
- `doc/api/cli.md`, Flag documentation
- `doc/api/index.md` / `doc/node.1`, Index updates
- `test/parallel/test-code-cache.js` / `test-process-get-builtin.mjs` / `test-require-resolve.js`, Test adjustments for new module

AugustinMauroy commented on lib/logger.js:
```diff
@@ -0,0 +1,383 @@
+  fatal: diagnosticsChannel.channel('log:fatal'),
+};
+
+// RFC5424 numerical ordering + log4j interface
```
```suggestion
/*
* RFC5424 numerical ordering + log4j interface
* @link https://www.rfc-editor.org/rfc/rfc5424.html
*/
```

so new comer can easily read the RFC

mertcanaltin commented on benchmark/logger/vs-pino.js:
```diff
@@ -0,0 +1,146 @@
+'use strict';
+
+const common = require('../common');
+const fs = require('fs');
```
thanks, I accepted

AugustinMauroy commented on lib/logger.js:
```diff
@@ -0,0 +1,386 @@
+  }
+}
+
+function createLogger(options) {
```
IMO we shouldn't expose this kind of API. now we didn't have this style of API

atlowChemi commented on lib/logger.js:
```diff
@@ -0,0 +1,393 @@
+    super(options);
+    const {
+      stream,
+      fields = {},
```
```suggestion
      fields = kEmptyObject,
```

atlowChemi commented on lib/logger.js:
```diff
@@ -0,0 +1,393 @@
+ * Logger class
+ */
+class Logger {
+  constructor(options = {}) {
```
```suggestion
  constructor(options = kEmptyObject) {
```

atlowChemi commented on lib/logger.js:
```diff
@@ -0,0 +1,393 @@
+    validateObject(options, 'options');
+    const {
+      level = 'info',
+      bindings = {},
```
```suggestion
      bindings = kEmptyObject,
```

atlowChemi commented on lib/logger.js:
```diff
@@ -0,0 +1,393 @@
+    if (!LEVELS[level]) {
+      throw new ERR_INVALID_ARG_VALUE('options.level', level,
+                                      `must be one of: ${LEVEL_NAMES.join(', ')}`);
+    }
```
Why not `validateOneOf` LEVEL_NAMES?

atlowChemi commented on lib/logger.js:
```diff
@@ -0,0 +1,393 @@
+   * @param {object} [options] - Optional overrides
+   * @returns {Logger}
+   */
+  child(bindings, options = {}) {
```
```suggestion
  child(bindings, options = kEmptyObject) {
```

atlowChemi commented on lib/logger.js:
```diff
@@ -0,0 +1,393 @@
+    return value !== null &&
+           typeof value === 'object' &&
+           (ObjectPrototypeToString(value) === '[object Error]' ||
+            value instanceof Error);
```
Why not `ErrorIsError`?

mertcanaltin replied:
I think this feature is not yet available in Node.js Core

atlowChemi replied:
> I think this feature is not yet available in Node.js Core

It is part of v8, and exists AFAIK starting from v24?

mertcanaltin replied:
you right!, thanks I finded

mertcanaltin replied:
https://github.com/nodejs/node/pull/60468/commits/1a76b7fa20a23ec51bc80cd0ebf262104dc53cb7

When I tried again it based on @marco-ippolito  suggestion, success. 🚀

mertcanaltin replied:
I was mistaken, https://github.com/nodejs/node/pull/60468#issuecomment-3611485415

I think ErrorIsError not defined in primordials,
I wonder your any suggestions? @marco-ippolito @atlowChemi

mertcanaltin replied:
I used isNativeError thanks @aduh95 (We talked on Slack)

atlowChemi commented on lib/logger.js:
```diff
@@ -0,0 +1,393 @@
+
+      if (logFields.err && this._isError(logFields.err)) {
+        logFields.err = this._serializeError(logFields.err);
+      }
```
Should we also support `logFields.error`?

atlowChemi commented on lib/logger.js:
```diff
@@ -0,0 +1,393 @@
+    }
+
+    return serialized;
+  }
```
Should we handle `cause` as well?

atlowChemi commented on lib/logger.js:
```diff
@@ -0,0 +1,393 @@
+}
+
+module.exports = {
+  Logger,
```
```suggestion
module.exports = {
  __proto__: null,
  Logger,
```

mcollina commented:
This will require support for serializers

mertcanaltin commented on test/parallel/test-log-basic.js:
```diff
@@ -0,0 +1,446 @@
+
+  // String without second arg should work (no assertion needed, just shouldn't throw)
+  logger.info('just a message');
+}// Test child logger context inheritance
```
I controlled but I view already new line in this file

mertcanaltin commented on lib/logger.js:
```diff
@@ -0,0 +1,393 @@
+      };
+    } else if (typeof msgOrObj === 'string') {
+      msg = msgOrObj;
+      logFields = fields || {};
```
I used nullish for null and undefined, thank you.

mertcanaltin commented on lib/logger.js:
```diff
@@ -0,0 +1,433 @@
+   * @param {object} record
+   * @private
+   */
+  _handleLog(level, record) {
```
thanks, I edited

mertcanaltin commented on lib/logger.js:
```diff
@@ -0,0 +1,433 @@
+
+    validateObject(fields, 'options.fields');
+
+    this._stream = stream ? this._createStream(stream) :
```
> Symbols are not truly private. Did you consider using private fields?

thanks for suggestion, yes I tryed now used private fields https://github.com/nodejs/node/pull/60468/commits/d94fd2bdd95b35f1ff1bf22427d1cc889acbab33

mertcanaltin commented:
> > > This will require support for serializers
> > 
> > 
> > I wonder, should we name them like Pino does (built-in serializers), or go with something like standardSerializers ?
> 
> Follow pino and we’ll change it

I tryed serializer implement for logger, and some bench result repaired

```
➜  node git:(mert/create-logger-api/node-core) ✗ ./node benchmark/logger/vs-pino.js n=100000
logger/vs-pino.js scenario="simple" logger="node-logger" n=100000: 4,868,164.032787085
logger/vs-pino.js scenario="child" logger="node-logger" n=100000: 3,894,327.425314102
logger/vs-pino.js scenario="disabled" logger="node-logger" n=100000: 160,503,080.85663706
logger/vs-pino.js scenario="fields" logger="node-logger" n=100000: 2,157,462.3927336666
logger/vs-pino.js scenario="simple" logger="pino" n=100000: 3,424,706.4418693925
logger/vs-pino.js scenario="child" logger="pino" n=100000: 4,753,595.477010947
logger/vs-pino.js scenario="disabled" logger="pino" n=100000: 122,100,122.10012211
logger/vs-pino.js scenario="fields" logger="pino" n=100000: 1,411,215.99189962
➜  node git:(mert/create-logger-api/node-core) ✗ 
```

mertcanaltin commented:
I will today inspect this error

```
Error: --- stderr ---
node:internal/logger/serializers:12
  if (!ErrorIsError(error)) {
       ^

TypeError: ErrorIsError is not a function
    at Object.serializeErr [as err] (node:internal/logger/serializers:12:8)
    at #log (node:logger:481:29)
    at Logger.error (node:logger:558:14)
    at Object.<anonymous> (/home/runner/work/node/node/node/test/parallel/test-log-basic.js:288:10)
    at Module._compile (node:internal/modules/cjs/loader:1760:14)
    at Object..js (node:internal/modules/cjs/loader:1891:10)
    at Module.load (node:internal/modules/cjs/loader:1480:32)
    at Module._load (node:internal/modules/cjs/loader:1299:12)
    at TracingChannel.traceSync (node:diagnostics_channel:328:14)
    at wrapModuleLoad (node:internal/modules/cjs/loader:245:24)

Node.js v26.0.0-pre
Command: out/Release/node /home/runner/work/node/node/node/test/parallel/test-log-basic.js
```

jsumners commented on lib/logger.js:
```diff
@@ -0,0 +1,584 @@
+   * @returns {boolean}
+   */
+  enabled(level) {
+    return LEVELS[level] >= this.#levelValue;
```
Please see how it is done in Pino: https://github.com/pinojs/pino/blob/731afd279ba9081c3032ef265c8943db9959ac90/lib/levels.js#L114-L118

Simply: if the level name doesn't exist, it can't be enabled. About the most handholding you can do here is to force name casing (`.toLowerCase`).

yvele replied:
@jsumners With all due respect, and if this function is intended to be used by Node.js end users, it might be worth double-checking with @marco-ippolito (security expert) or @atlowChemi.

Explicit input validation can sometimes be a security concern, but it also helps make code more predictable and safer for end users. This is highlighted in [OWASP Top 10 - Mishandling of Exceptional Conditions (A10, 2025)](https://owasp.org/Top10/2025/A10_2025-Mishandling_of_Exceptional_Conditions/) (silent failure).

I’m not a Node.js core expert, so I defer to your judgment, just wanted to flag this as a potential area for review.

This will be my last note on this, I won’t push further if you decide it’s not worth addressing 😉

mertcanaltin replied:
Maybe mentioning this in the doc can guide the user in the right direction and alleviate their concerns @yvele

jsumners replied:
I greatly regret adding custom levels to Pino. I strongly recommend skipping them here.

mertcanaltin replied:
Thank you both for the discussion, I'll keep the current behavior where [enabled()](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) returns false for unknown levels. I agree with @jsumners that we shouldn't add custom levels, and the current implementation is simple and safe. I'll document this behavior clearly in the API docs https://github.com/nodejs/node/pull/60468/commits/83a12b70258efd8c58d33e92d135b4d1560e9f2f.

mertcanaltin replied:
thanks, yes this topic solved

mertcanaltin commented:
I tried running `parallel/test-debugger-backtrace`  test file locally and succes

mertcanaltin commented on test/parallel/test-log-basic.js:
```diff
@@ -408,24 +407,23 @@ const path = require('node:path');
   const lines2 = output2.trim().split('\n');
 
-  // Consumer1 should have all 4 logs (debug+)
+  // Consumer1 dosya formatı satır başına log yazıyorsa 4 olmalı, değilse bu kontrolü güncelle.
```
thanks for ping

mcollina commented on doc/api/logger.md:
```diff
@@ -0,0 +1,40 @@
+# Logger
```
this file is really empty

mcollina commented on lib/logger.js:
```diff
@@ -0,0 +1,589 @@
+   * @returns {object} Serialized error object
+   * @private
+   */
+  #serializeError(err, seen = new SafeSet()) {
```
I think this should be in a serializer

mertcanaltin replied:
Yes, I was added in a serialize, https://github.com/nodejs/node/pull/60468/changes#diff-2d1249526faad148f37dc1ea3bfd8d56b77ceb4bf1e27299de1ed8e1d6ffc12cR9

anonrig commented on lib/internal/logger/serializers.js:
```diff
@@ -0,0 +1,73 @@
+  }
+
+  const obj = {
+    type: error.constructor.name,
```
```suggestion
  const obj = {
    __proto__: null,
    type: error.constructor.name,
```

anonrig commented on lib/internal/logger/serializers.js:
```diff
@@ -0,0 +1,73 @@
+
+/**
+ * Serializes HTTP request object
+ * @param {object} req - HTTP request
```
Can you extend the jsdoc here? Something like this, but with better indentation:

```suggestion
 * @param {{method: string, url: string, headers: string, socket: undefined | {remoteAddress: string, remotePort: string}} req - HTTP request
```

anonrig commented on lib/logger.js:
```diff
@@ -0,0 +1,589 @@
+const stdSerializers = require('internal/logger/serializers');
+
+// Create channels for each log level
+const channels = {
```
```suggestion
const channels = {
  __proto__: null,
```

anonrig commented on lib/logger.js:
```diff
@@ -0,0 +1,589 @@
+ * Consumers subscribe to diagnostics_channel events
+ */
+class LogConsumer {
+  #levelValue;
```
```suggestion
  /** @type {number} */
  #levelValue;
```

anonrig commented on lib/logger.js:
```diff
@@ -0,0 +1,589 @@
+  #stream;
+  #fields;
+
+  constructor(options = kEmptyObject) {
```
Can you add JSDoc to all methods with proper types? It's hard to follow up on the parameters and what options are supported etc.

mertcanaltin replied:
Thanks for all suggestions, I tryed edited @anonrig

anonrig commented on lib/logger.js:
```diff
@@ -0,0 +1,393 @@
+    if (!LEVELS[level]) {
+      throw new ERR_INVALID_ARG_VALUE('options.level', level,
+                                      `must be one of: ${LEVEL_NAMES.join(', ')}`);
+    }
```
This still doesn't use validateOneOf

mertcanaltin commented on lib/internal/logger/serializers.js:
```diff
@@ -0,0 +1,73 @@
+'use strict';
+const { isNativeError } = require('internal/util/types');
```
Unfortunately, no. I'm trying to create one, but the first problem I encountered is that Benchmark I understand it's not possible to run it directly using internal modules. I'll look for a different solution,
Do you have any suggestions? @anonrig

Qard commented on lib/logger.js:
```diff
@@ -0,0 +1,625 @@
+      }
+    }
+
+    const childLogger = new Logger({
```
Do we need `__proto__: null` here?

Qard commented on test/parallel/test-logger-serializers.js:
```diff
@@ -0,0 +1,194 @@
+  const log = stream.logs[0];
+  assert.strictEqual(log.req.method, 'GET');
+  assert.strictEqual(log.req.url, '/test');
+  assert.strictEqual(log.res.statusCode, 200);
```
Should validate exclusion of things we _don't_ want passing through.

Qard reviewed (commented):
I think the general direction is good. I mostly have a few internal detail concerns. For API surface level, I think it's generally fine, but I'm not a huge fan of the `logger.enabled(arbitraryString)` thing and I'm also a bit confused by having level reconfigurable per-child. Why is that exactly? I'm unclear on the value of having independent log levels per child.

TheOneTheOnlyJJ commented:
> Or should we keep both approaches?

I guess the arbitrary string method (as currently implemented) makes sense if there can be new log levels added at run-time.

> Independent log levels per child

This is a nice feature to have, and it would be unfortunate to have it removed now that it is already implemented. But I am not a maintainer, so I will not comment on it more than this.

mertcanaltin commented:
Thanks for comments @TheOneTheOnlyJJ & @Qard

  Implemented logger.<level>.enabled getter https://github.com/nodejs/node/pull/60468/commits/ad43d259fd16a97a40f6ffb10425da557fa49217:

  ```js
  if (logger.debug.enabled) {
    logger.debug('expensive', computeData());
  }
  // Typos throw: logger.debg.enabled → TypeError

  For per-child levels, keeping it as-is with clear docs - the use cases (module-specific debugging) seem valuable enough.
  ```

avivkeller commented on lib/logger.js:
```diff
@@ -0,0 +1,659 @@
+    // Start building JSON manually
+    let json = '{"level":"' + record.level + '",' +
+               '"time":' + record.time + ',' +
+               '"msg":"' + record.msg + '"';
```
Why aren't we building the JSON like an object, and then using `JSONStringify` on the entire object? I feel like it'll be a lot more readable.

mertcanaltin replied:
we used manual concat for performance because JSONStringify slowly

avivkeller commented on lib/logger.js:
```diff
@@ -0,0 +1,659 @@
+  #log(level, levelValue, msgOrObj, fields) {
+    if (levelValue < this.#levelValue) {
+      return;
+    }
```
Don't we have a `this.#enabled(level)` helper?

mertcanaltin commented:
> I thought a bit more about the per-child log level thing. Do you think it would be reasonable to suggest in docs that module authors should NOT use it in their own code and that it should generally only be used in application code for isolating particular components or direct dependencies?
> 
> I'm just thinking it would be annoying to have a transitive dependency not follow the log levels an application developer specifies. I would think the typical pattern with loggers, at least as I have seen before, would be to construct a child externally to the library you want to use the logger, that way you can control the behaviour external to that module? But then with a library calling another library it may construct another child but should just inherit the logger level it has been given by the application developer?
> 
> What do you think? Should we document the feature as something generally only application code should use?

Yes, I agree. I added this note to the doc thanks for idea. https://github.com/nodejs/node/pull/60468/commits/0cc656b216cede1c3476e2fcf232a40db43dc7dc

mertcanaltin commented on lib/logger.js:
```diff
@@ -0,0 +1,659 @@
+   * @private
+   */
+  #createLogMethod(level, methodLevelValue, loggerLevelValue) {
+    const enabled = methodLevelValue >= loggerLevelValue;
```
the public enabled() method performs validation for unknown levels. here we skip that since we're passing known valid levels internally, it's a performance optimization.

mertcanaltin commented on lib/logger.js:
```diff
@@ -0,0 +1,659 @@
+    const { level = 'info' } = options;
+    validateString(level, 'options.level');
+
+    validateOneOf(level, 'options.level', LEVEL_NAMES);
```
thanks, I edited

mertcanaltin commented:
If you need to change log levels at runtime, you can create a new child logger:

```js
// Start with info level
const logger = new Logger({ level: 'info' });

// Later, when you need debug logging
const debugLogger = logger.child({}, { level: 'debug' });
  ```

TheOneTheOnlyJJ commented:
I guess the final decision will be based on whether minimizing external dependencies for logging is a (secondary) goal of this API or not.

mertcanaltin commented on lib/logger.js:
```diff
@@ -0,0 +1,601 @@
+
+  handle(record) {
+    // Start building JSON manually
+    let json = '{"level":"' + record.level + '",' +
```
nice catch!, template literal is very clean, thanks

mertcanaltin commented:
> Imho this module should only be available behind a flag `--experimental-logger`

+1  I added  `--experimental-logger` flag

mertcanaltin commented on lib/internal/logger/serializers.js:
```diff
@@ -0,0 +1,76 @@
+ * } }} req - HTTP request
+ * @returns {{ method: string, url: string, headers: object, remoteAddress?: string, remotePort?: string }}
+ */
+function req(req) {
```
I tryed now https://github.com/nodejs/node/pull/60468/commits/b53d408772efd006f94661bd0aea1cf7ed1760b0

Qard commented on doc/api/logger.md:
```diff
@@ -0,0 +1,839 @@
+
+* `msg` {string} Log message.
+* `obj` {Object} Object containing `msg` property and additional fields.
+* `error` {Error} Error object to log.
```
It feels slightly weird to me that errors are accepted in _all_ the interfaces rather than just the `error(...)` interface specifically, but I guess the functions are treated as being homogenous so... 🤷🏻

Not an issue, just an observation. Undecided if I think we should just not document this? Or if we _are_ going to document it should we document it _more_ with actual examples for each function?

Qard commented on doc/api/logger.md:
````diff
@@ -0,0 +1,839 @@
+consumer.end();
+```
+
+## `logger.stdSerializers`
````
If we have the `[serialize]()` thing, can we not just set those on the error, request, and response types directly in core rather than needing to export and use these serializer functions here?

Qard commented on lib/logger.js:
```diff
@@ -0,0 +1,594 @@
+/**
+ * JSON consumer - outputs structured JSON logs
+ */
+class JSONConsumer extends LogConsumer {
```
I wonder if we should have a `Utf8StreamConsumer` class between these two in order to make all that stream handling stuff reusable for other custom consumers. 🤔

Qard commented on lib/logger.js:
```diff
@@ -0,0 +1,594 @@
+    validateObject(serializers, 'options.serializers');
+
+    // Validate serializers are functions
+    for (const key in serializers) {
```
Generally best to avoid for-in loops on user input as it will walk the entire prototype tree. This would pull things out of parent classes of the inputs to try and validate as functions. This is also how prototype injection attacks often happen.

mertcanaltin commented on doc/api/logger.md:
````diff
@@ -0,0 +1,839 @@
+channels.info.subscribe((record) => {
+  // Custom handling of info-level logs
+});
+```
````
Noted. Since strong coupling isn't a concern in core, keeping it for now. Can revisit if needed.

mertcanaltin commented on lib/logger.js:
```diff
@@ -0,0 +1,594 @@
+    let result = '';
+    for (const key in bindings) {
+      const serialized = this.#serializeValue(bindings[key], key);
+      result += `,"${key}":${JSONStringify(serialized)}`;
```
Yes, bindings are pre-serialized to JSON for performance (avoids re-serializing on every log call). Trade-off: fasterlogging but JSON-only format.

AugustinMauroy commented on doc/api/logger.md:
````diff
@@ -0,0 +1,818 @@
+
+### Custom consumer
+
+```mjs
````
this should use `node:util` with `styleText` 😁

AugustinMauroy reviewed (approved):
LGTM !

RafaelGSS commented on benchmark/logger/vs-pino.js:
```diff
@@ -0,0 +1,146 @@
+'use strict';
```
I don't think we should have this benchmark on core. A separate repo using the custom binary might be enough.

RafaelGSS commented on test/es-module/test-require-module-transpiled.js:
```diff
@@ -1,30 +1,36 @@
 
-spawnSyncAndAssert(process.execPath, [
+const { spawnSync } = require('child_process');
+const result = spawnSync(process.execPath, [
```
I'm pretty sure none of this should be changed on this PR.

mertcanaltin commented:
@AugustinMauroy @RafaelGSS Thanks for the review, edited your comments

RafaelGSS commented on doc/api/logger.md:
```diff
@@ -0,0 +1,819 @@
+
+<!--introduced_in=v26.0.0-->
+
+> Stability: 1.0 - Early Development
```
```suggestion
> Stability: 1.1 - Active Development
```

RafaelGSS commented on doc/api/logger.md:
```diff
@@ -0,0 +1,819 @@
+* `msg` {string} Log message.
+* `obj` {Object} Object containing `msg` property and additional fields.
+* `error` {Error} Error object to log.
+* `fields` {Object} Additional fields to include in the log record.
```
It's not clear its impact on the call

RafaelGSS commented on doc/api/logger.md:
```diff
@@ -0,0 +1,819 @@
+* `msg` {string} Log message.
+* `obj` {Object} Object containing `msg` property and additional fields.
+* `error` {Error} Error object to log.
+* `fields` {Object} Additional fields to include in the log record.
```
Ditto

RafaelGSS commented on lib/internal/logger/serializers.js:
```diff
@@ -0,0 +1,77 @@
+ */
+function serializeErr(error) {
+  if (!isNativeError(error)) {
+    return error;
```
It's not clear to me why we return if it's not a native error; shouldn't we log/fail instead?

RafaelGSS commented on lib/logger.js:
```diff
@@ -0,0 +1,587 @@
+// Create channels for each log level
+const channels = {
+  __proto__: null,
+  trace: diagnosticsChannel.channel('log:trace'),
```
we should document the channel names used somewhere.

RafaelGSS reviewed (changes requested):
It's missing tests for a significant part of the exposed API. Apart from my comments, why are we copying `pino` internals instead of vendoring `pino` (which is more robust and well tested) into core? 

I'd love to hear @mcollina and @jsumners thoughts on this PR too.

RafaelGSS commented:
> @RafaelGSS please take a look. 7 days passed since your review has been requested.

It's a 2.5k lines PR, it's hard to concilate time to review it properly.

mertcanaltin committed:
lib: add logger api in node core

Signed-off-by: Mert Can Altin <mertgold60@gmail.com>

mertcanaltin committed:
lib: address review feedback for logger API

Signed-off-by: Mert Can Altin <mertgold60@gmail.com>

mertcanaltin committed:
lib: address review feedback for logger API

Signed-off-by: Mert Can Altin <mertgold60@gmail.com>

mertcanaltin committed:
added missing test

Signed-off-by: Mert Can Altin <mertgold60@gmail.com>

H4ad commented on lib/logger.js:
```diff
@@ -0,0 +1,587 @@
+
+  handle(record) {
+    // Start building JSON manually
+    let json = `{"level":"${record.level}","time":${record.time},"msg":"${record.msg}"`;
```
We should also add validation for fields we don't trust, like record.msg

H4ad commented on test/parallel/test-logger-serializers.js:
We should add tests with invalid data: `logger.info('My invalid json ", "__proto__": true,')`,

mertcanaltin replied:
thanks, added

H4ad commented on lib/logger.js:
```diff
@@ -0,0 +1,587 @@
+        channels[level].subscribe(this.handle.bind(this));
+      }
+    }
+  }
```
Why no detach?

mertcanaltin commented on lib/logger.js:
```diff
@@ -0,0 +1,587 @@
+
+    return serialized;
+  }
+}
```
this is by design, [serialize]() lets the object control what it exposes (e.g. strip sensitive fields), then the field serializer lets the logger transform the result (e.g. add metadata). Two different concerns, intentionally stacked. Same approach as pino.

mertcanaltin committed:
logger: address review feedback

Signed-off-by: Mert Can Altin <mertgold60@gmail.com>

mertcanaltin committed:
logger: address review feedback

Signed-off-by: Mert Can Altin <mertgold60@gmail.com>

mertcanaltin commented on lib/logger.js:
```diff
@@ -0,0 +1,587 @@
+      const value = consumerFields[key];
+      json += typeof value === 'string' ?
+        `,"${key}":"${value}"` :
+        `,"${key}":${JSONStringify(value)}`;
```
thanks!

mertcanaltin commented on lib/logger.js:
```diff
@@ -0,0 +1,587 @@
+
+    // Step 2: Apply field-specific serializer, if present
+    if (this.#serializers[key]) {
+      result = this.#serializers[key](result);
```
This is by design, serialize lets the object control what it exposes (e.g. strip sensitive fields), then the field serializer lets the logger transform the result (e.g. add metadata). Two different concerns, intentionally stacked. Same approach as pino.

mcollina commented:
On vedoring pino in core: IMHO it’s too opinionated to be in core. It works for me, but it uses some historical/non-standard conventions that makes it hard for some users to use it.

An option we could follow is to follow the same path of undici and create something new with the goal of bringing it in the fold.

mertcanaltin committed:
Update lib/internal/logger/serializers.js

Co-authored-by: Yagiz Nizipli <yagiz@nizipli.com>

Signed-off-by: Mert Can Altin <mertgold60@gmail.com>

mertcanaltin commented on lib/internal/logger/serializers.js:
```diff
@@ -0,0 +1,89 @@
+  // Circular reference protection
+  if (seen === undefined) {
+    seen = new SafeSet();
+  }
```
thanks

mertcanaltin committed:
delete pino benchmark

Signed-off-by: Mert Can Altin <mertgold60@gmail.com>

mertcanaltin commented on lib/logger.js:
```diff
@@ -0,0 +1,548 @@
+  channels,
+  stdSerializers,
+  serialize,
+};
```
Okey

mertcanaltin committed:
delete exports (LEVELS, channels)

Signed-off-by: Mert Can Altin <mertgold60@gmail.com>

mertcanaltin commented on doc/api/logger.md:
```diff
@@ -0,0 +1,825 @@
+The `node:logger` module provides high-performance structured logging
+capabilities for Node.js applications. It uses `diagnostics_channel` internally
+to dispatch log events to consumers, allowing multiple consumers to receive
+logs independently.
```
I agree, I’ll send a commit with the changes shortly.

jasnell reviewed (dismissed):
I'd prefer `require('console/logging')` or `require('console').Logging` rather than a new top-level `node:...` module.

mertcanaltin committed:
edited logger api md

Signed-off-by: Mert Can Altin <mertgold60@gmail.com>

mertcanaltin committed:
address review feedback for logger api

Signed-off-by: Mert Can Altin <mertgold60@gmail.com>

mertcanaltin commented on lib/internal/logger/serializers.js:
```diff
@@ -0,0 +1,87 @@
+  // Handle error cause recursively
+  if (error.cause !== undefined) {
+    obj.cause = typeof error.cause === 'object' && error.cause !== null ?
+      serializeErr(error.cause, seen) :
```
Thanks, switched to iterative while loop with a ```serializeOneError``` helper.

mertcanaltin commented on lib/logger.js:
```diff
@@ -0,0 +1,546 @@
+
+    // Object: custom stream with write method
+    if (typeof stream === 'object' && stream !== null &&
+        typeof stream.write === 'function') {
```
Thanks, simplified to t```ypeof stream?.write === 'function'```.

mertcanaltin commented on lib/logger.js:
```diff
@@ -0,0 +1,546 @@
+   * Close the consumer
+   */
+  end() {
+    this.#stream.end();
```
Yes, nulled out, also added ```this.detach()``` ```before end()``` to prevent null-dereference if consumer is still attached.

mertcanaltin committed:
doc: address review feedback on logger.md

Signed-off-by: Mert Can Altin <mertgold60@gmail.com>

mertcanaltin committed:
doc: add log channels in diagnostics_channel

Signed-off-by: Mert Can Altin <mertgold60@gmail.com>

mertcanaltin committed:
logger: skip input validation when no subscribers

Signed-off-by: Mert Can Altin <mertgold60@gmail.com>

mertcanaltin commented:
Hello everyone, I pushed a small follow-up, moved the hasSubscribers check before input validation so disabled consumers skip parsing (the original intent from @atlowChemi / @Qard).

Plus a few doc bits:
Error exception note in the JSON section, log channel names under LogConsumer and in
```diagnostics_channel.md```, and a Stability 1.1 marker on ```logger.child()```

@RafaelGSS this round also covers your remaining points (channel names + Stability 1.1 on child), mind taking another look when you get a chance?

Thanks again for all the reviews.

atlowChemi commented on lib/internal/logger/serializers.js:
```diff
@@ -0,0 +1,93 @@
+ * @returns {{ method: string, url: string, headers: object, remoteAddress?: string, remotePort?: string }}
+ */
+function req(req) {
+  return {
```
```suggestion
  return {
    __proto__: null,
```

mertcanaltin replied:
thanks for edit

atlowChemi commented on lib/internal/logger/serializers.js:
```diff
@@ -0,0 +1,93 @@
+ * @returns {{ statusCode: number, headers: object }}
+ */
+function res(res) {
+  return {
```
```suggestion
  return {
    __proto__: null,
```

atlowChemi commented on lib/internal/logger/serializers.js:
```diff
@@ -0,0 +1,93 @@
+  };
+}
+
+module.exports = {
```
```suggestion
module.exports = {
  __proto__: null,
```

atlowChemi commented on lib/logger.js:
```diff
@@ -0,0 +1,548 @@
+      }
+    }
+
+    const record = {
```
```suggestion
    const record = {
      __proto__: null,
```

mertcanaltin replied:
thanks!

mertcanaltin committed:
Update serializers.js

Co-authored-by: Chemi Atlow <chemi@atlow.co.il>

Signed-off-by: Mert Can Altin <mertgold60@gmail.com>

mertcanaltin committed:
Update serializers.js

Co-authored-by: Chemi Atlow <chemi@atlow.co.il>

Signed-off-by: Mert Can Altin <mertgold60@gmail.com>

mertcanaltin committed:
Update logger.js

Co-authored-by: Chemi Atlow <chemi@atlow.co.il>

Signed-off-by: Mert Can Altin <mertgold60@gmail.com>

mertcanaltin committed:
Update lib/internal/logger/serializers.js

Co-authored-by: Chemi Atlow <chemi@atlow.co.il>

Signed-off-by: Mert Can Altin <mertgold60@gmail.com>

mertcanaltin commented:
https://github.com/nodejs/node/actions/runs/24440675298/job/71405021240?pr=60468
https://github.com/nodejs/node/actions/runs/24440675298/job/71405021236?pr=60468

I looked this error, I guess sometimes this is a flaky test, I continue inspecting

https://github.com/nodejs/node/issues/53579

mertcanaltin commented:
Hello again I try  this two test in my linux machine and get a passed

```shell
mert@lima-default:/Users/mert/Desktop/enjoy/linux-node/node$ out/Release/node test/parallel/test-snapshot-reproducible.js
mert@lima-default:/Users/mert/Desktop/enjoy/linux-node/node$ 
```

mertcanaltin committed:
Merge branch 'main' into mert/create-logger-api/node-core

Signed-off-by: Mert Can Altin <mertgold60@gmail.com>

H4ad commented on lib/logger.js:
```diff
@@ -0,0 +1,549 @@
+    // Add consumer fields
+    const consumerFields = this.#fields;
+    for (const key of ObjectKeys(consumerFields)) {
+      json += `,${JSONStringify(key)}:${JSONStringify(consumerFields[key])}`;
```
We need a if check to ensure consumerFIelds[key] is not undefined, otherwise will produce invalid JSON.

H4ad replied:
And same question as the other, what we do when fields has keys conflicting with other fields?

mertcanaltin replied:
Fixed, skipping undefined in handle() too.

mertcanaltin replied:
> And same question as the other, what we do when fields has keys conflicting with other fields?

By design, the last used value takes precedence (can you check the field override priority test?)
The `level` `time` `msg` fields are reserved.

H4ad commented on lib/logger.js:
```diff
@@ -0,0 +1,549 @@
+    // Add log fields
+    const fields = record.fields;
+    for (const key in fields) {
+      json += `,${JSONStringify(key)}:${JSONStringify(fields[key])}`;
```
Same for this, check for fields[key] undefined

H4ad commented on lib/logger.js:
```diff
@@ -0,0 +1,549 @@
+
+    let result = '';
+    const keys = ObjectKeys(bindings);
+    for (const key of keys) {
+      const serialized = this.#serializeValue(bindings[key], key);
+      result += `,${JSONStringify(key)}:${JSONStringify(serialized)}`;
```
Same for this, check for undefined to avoid invalid JSON.

mertcanaltin replied:
Skipping undefined in #serializeBindings too.

H4ad commented on lib/logger.js:
```diff
@@ -0,0 +1,549 @@
+    const levelValue = this.#levelValue;
+    for (let i = 0; i < LEVEL_NAMES.length; i++) {
+      const lvl = LEVEL_NAMES[i];
+      this[lvl] = { __proto__: null, enabled: LEVELS[lvl] >= levelValue };
+    }
+  }
+
+  /**
+   * Attach this consumer to log channels
+   */
+  attach() {
+    for (const level of LEVEL_NAMES) {
+      if (this[level].enabled) {
+        const handler = this.handle.bind(this);
+        this.#handlers[level] = handler;
+        channels[level].subscribe(handler);
+      }
+    }
+  }
```
Add flag to ensure we can't attach twice

mertcanaltin replied:
Added #attached flag, attach are idempotent now.

H4ad commented on doc/api/logger.md:
```diff
@@ -0,0 +1,920 @@
+* `options` {Object}
+  * `level` {string} Minimum log level to consume. **Default:** `'info'`.
+  * `stream` {number|string|Object} Output destination. Can be a file
+    descriptor (number), file path (string), or a writable stream object.
```
Would not be better to create an interface for `object` and explicitly declare the methods allowed? Eg: JSONConsumer calls flush/flushSync so I can't use it with createWriteStream

mertcanaltin replied:
Now requires write, flush, flushSync, end, docs list them.

H4ad commented on lib/logger.js:
```diff
@@ -0,0 +1,549 @@
+      };
+
+      // Apply serializers to additional fields
+      if (fields) {
```
missing validate fields as object when is defined.

mertcanaltin replied:
Hoisted validateObject(fields, 'fields') so it runs for all branches.

H4ad commented on lib/logger.js:
```diff
@@ -0,0 +1,549 @@
+    let logFields;
+
+    if (isNativeError(msgOrObj)) {
+      msg = msgOrObj.message;
```
If this is undefined, it will likely produce invalid json

mertcanaltin replied:
Added ?? '' fallback for subclasses that leave message undefined.

H4ad commented on lib/logger.js:
```diff
@@ -0,0 +1,549 @@
+      logFields = fields ? this.#applySerializers(fields) : kEmptyObject;
+    } else {
+      const { msg: extractedMsg, ...restFields } = msgOrObj;
+      msg = extractedMsg;
```
Same for this one, can produce invalid json

mertcanaltin replied:
Already guarded by validateString(msgOrObj.msg) above.

H4ad commented on lib/logger.js:
```diff
@@ -0,0 +1,549 @@
+    for (const key of ObjectKeys(consumerFields)) {
+      json += `,${JSONStringify(key)}:${JSONStringify(consumerFields[key])}`;
+    }
+
+    // Add pre-serialized bindings
+    json += record.bindingsStr;
+
+    // Add log fields
+    const fields = record.fields;
+    for (const key in fields) {
+      json += `,${JSONStringify(key)}:${JSONStringify(fields[key])}`;
+    }
```
Why not serialize the entire fields and just remove first/last bracket?

What happens when fields has keys that conflicts with other fields?

mertcanaltin replied:
Bulk stringify: follow-up with a bench.

ast-wins by design, level, time, msg reserved.

mertcanaltin committed:
logger: address review feedback

Signed-off-by: Mert Can Altin <mertgold60@gmail.com>

mertcanaltin committed:
Merge branch 'main' into mert/create-logger-api/node-core

Signed-off-by: Mert Can Altin <mertgold60@gmail.com>

mertcanaltin committed:
src: remove duplicate experimental_stream_iter

Signed-off-by: Mert Can Altin <mertgold60@gmail.com>

mcollina commented:
@anonrig what is the question for the TSC?

mcollina commented:
No question for the TSC, removing.

H4ad commented:
@mcollina the question was not related to the module this logger API would be released?

> I didn't see the other comment, sorry for not sending this earlier.

mcollina commented:
@H4ad I don't understand.

H4ad commented:
@mcollina If the new logger will be released as `node:logger` or `node:console/logging`

Ref:
- https://github.com/nodejs/node/pull/60468#discussion_r3006491428
- https://github.com/nodejs/node/pull/60468#pullrequestreview-4027054601

mertcanaltin committed:
Merge branch 'main' into mert/create-logger-api/node-core

Signed-off-by: Mert Can Altin <mertgold60@gmail.com>

mcollina commented on doc/api/logger.md:
````diff
@@ -0,0 +1,825 @@
+logs independently.
+
+```mjs
+import { Logger, JSONConsumer } from 'node:logger';
````
> Adding a new module should be a minor change

IMHO this was due to the "shadowing" of the ecosystem. I think this is relaxable/not needed for `node:` prefixed modules.

mcollina reviewed (changes requested):
I have a few issues with this implementation, which probably would need a bit more design and community discussion before landing.

### the diagnostics channels are global

The main problem with this is that the diagnostic channels are global. This match reality for most applications I've seen where logging is a "global" concern, however it makes it harder for testing. We create a logger with `new`, which implies a local scope, even though the logs go to a global stream.

Let's say that I what to create _two or more instances_ of the logger with two different transports, how can I achieve that?

(I still think that diagnostics_channel are the best primitive here, but we need to have answers to these questions?

### levels are strings, not numbers, in the diagnostics channels.

In most implementations, log levels are actually mapped to integers. This has an advantage when further processing them down. We can see this in action with the _n_ number of diagnostic channel listeners. This is present here, but not moved down the diagnostics channels. Why do we have one channel per level instead of just a single channel with integer values?

The concept of "ERROR is more important than INFO" is pushed down to consumers, and not built into the lib.

As an additional question: many companies want to create their own custom levels (unfortunately), how can this be implemented here?

Note this is about tradeoffs.

### the basic logger has a pre-serialized binding string

Part of the goal of this library is to be protocol agnostic, however there is a `#bindingStr`, which is some inheritance from Pino. I don't think we can afford to do this here in a generic way.
Child logger properties should be LRUd/cached down in the consumer.

mcollina commented:
> @mcollina If the new logger will be released as `node:logger` or `node:console/logging`

My understanding is there are quite a few plus in doing `node:logger` vs `node:console/logging`.

mertcanaltin commented:
>how can I achieve that?

right, unfortunately current desing is acting wrongly, I tried create new consumer but hit the some global key "log:info"
I think the fix is fewer channels, not more, a single channel with a namespace per logger.

> the basic logger has a pre-serialized binding string
> Child logger properties should be LRUd/cached down in the consumer.

yes not protocol agnostic current code, if as you said this different consumer, example msgpack or send to opentelemetry the current record only carries a JSON string, so a non-JSON consumer can't use it

but if I use protocol-agnostic, it work for all consumer, but without cache it serialize every log, so million times

I think good weakMap for this problem, fast and protocol agnostic

>levels are strings, not numbers

sgtm, as you said, the level is already in the record, so we can use a single channel and send the integer level in the payload, since the level is just a number, custom levels become easy, without the lib knowing it in advence

mertcanaltin committed:
use single diagnostics channel with consumer-side filtering

Signed-off-by: Mert Can Altin <mertgold60@gmail.com>

mertcanaltin commented:
Hello @mcollina, I sent a repair commit for your comments, now we use a single diagnostics channel, bindings cache  as you said moved to consumer, removed before serialized bindingsStr , now has a numeric level for all record

mertcanaltin committed:
updated doc

mertcanaltin commented:
I want to get your comments on custom levels each record already has levelValue, so no core changes needed to add them later

mertcanaltin committed:
apply serializers to bindings in Logger constructor

mertcanaltin committed:
remove double serialize

mertcanaltin committed:
simplify bindings handling in child logger creation

mertcanaltin committed:
add default error serializers to logger class

mertcanaltin commented:
Hi @mcollina, I've addressed your review comments in [2e64414](https://github.com/nodejs/node/pull/60468/commits/2e64414ce7666480d1080f98c1efb5811399b483). The logger now publishes to a single log channel, and consumers filter by level on their side.

mertcanaltin commented:
I solved the conflict.

mertcanaltin commented:
To avoid taking up any more of the reviewers time, I'm closing this PR. Thanks a lot, everyone. If anyone wants to pick this up and needs help, I'd be happy to assist them.
